### PR TITLE
Add tld cache extra prefixes to zipper

### PR DESCRIPTION
**Pull request by emadolsky**
## What issue is this change attempting to solve?
This new config adds the functionality to probe deeper paths on zipper in order to involve less backends on queries when metrics are broken into different cluster from the non-leading namespaces.

## How does this change solve the problem? Why is this the best approach?
In addition to normal TLD probes, probes with the configured prefixes are also sent. the results already gathered from the previous probes sent at the same iteration is used to optimize the backends receiving the probes.

## How can we be sure this works as expected?
New tests are introduced and the functionality is manually tested on local.

### CODE REVIEW

File: pkg/app/zipper/tldcache.go
Comment (timtofan) [inline pkg/app/zipper/tldcache.go:24]: imho it's better to never override loop variable for better readability

It will work, but the one who reads the code needs to know that range copies the value, and "original" value won't be changed. I think depending on knowing such details doesn't help readability.

File: pkg/app/zipper/tldcache.go
Comment (timtofan) [inline pkg/app/zipper/tldcache.go:59]: nit: imho it would be simpler to calculate nsCount first, and have single code path for assignment

File: pkg/app/zipper/tldcache.go
Comment (timtofan) [inline pkg/app/zipper/tldcache.go:91]: imho we shouldn't be modifying input arg, but have a separate variable for mutation so that it's easier to see lifecycle of the variable

File: pkg/app/zipper/tldcache.go
Comment (timtofan) [inline pkg/app/zipper/tldcache.go:79]: imho eternal loop is a dangerous construct, and generally should be avoided

maybe we could structure the code w/o it 🤔

File: pkg/app/zipper/tldcache.go
Comment (timtofan) [inline pkg/app/zipper/tldcache.go:82]: maybe those 2 lines can be replaced by return 🤔

File: pkg/app/zipper/tldcache.go
Comment (timtofan) [inline pkg/app/zipper/tldcache.go:107]: nit: we could also do it like this
```
	request := types.NewFindRequest(strings.join([]string{prefix,"*"}, ".")
```

dunno which approach is clearer though 🤔

File: pkg/app/zipper/tldcache.go
Comment (timtofan) [inline pkg/app/zipper/tldcache.go:77]: Imho this func is doing something unexpected. It's called getBackendsForPrefix, yet when there is no match it's not returning empty array. Maybe getBackendsForPrefixOrAll might be more intuitive 🤔

File: pkg/app/zipper/tldcache.go
Comment (timtofan) [inline pkg/app/zipper/tldcache.go:26]: nit: maybe we could use range() here? it's a tad simpler and is kinda default usually

File: pkg/app/zipper/tldcache.go
Comment (timtofan) [inline pkg/app/zipper/tldcache.go:27]: it's not introduced by you, but since you are touching that code:
I think it's a little weird to use longer topLevelDomain all over the place given we already use TLD abbreviation in package name and other places. IMHO we should be consistent in using either short or long version. And in this particular case, imho short version would be preferred since it's already established acronym in config and package name.

File: pkg/app/zipper/tldcache.go
Comment (emadolsky) [reply pkg/app/zipper/tldcache.go:103]: actually this ends up creating a string starting with "." if prefix is "". So I thought the condition was needed.

Comment (timtofan) [reply pkg/app/zipper/tldcache.go:103]: ah 🤦

File: pkg/app/zipper/tldcache.go
Comment (timtofan) [inline pkg/app/zipper/tldcache.go:76]: nit: if it's formatted as proper sentences w/ dots, then might be a good idea to also start the sentences from capital letter

File: pkg/cfg/common.go
Comment (grzkv) [inline pkg/cfg/common.go:182]: If I understand correctly, we can add prefixes that have more than one level of hierarchy. For example, we can have `a.b.c`, `a.b.d`, and `hello` here.

We may consider using a tree of strings here instead of a list. So, we will go from:

```
tldCacheExtraPrefixes:
  - "a.b.c"
  - "a.b.d"
  - "hello"
```
to for example
```
tldCacheExtraPrefixes:
  - "a":
      "b":
        - "c"
        - "d"
  - "hello"
```
(of course, there are also other ways to specify hierarchy, like with a nested list).

This will be more type-safe. It would make parsing a bit harder but then the other parsing and sorting logic will become much simpler.

At the same time, we do not expect many levels of hierarchy. So, it will still be convenient when we want to have e.g. a single prefix `a.b.c`.

File: pkg/app/zipper/tldcache.go
Comment (grzkv) [inline pkg/app/zipper/tldcache.go:92]: A minor thing. This seems like a proxy function. Please consider removing it.

File: pkg/app/zipper/tldcache.go
Comment (grzkv) [inline pkg/app/zipper/tldcache.go:22]: We pre-process the prefixes configuration - sorting, trimming, etc. - every time we probe the TLDs. If I'm not mistaken, we don't need to do this repeatedly. We only need to do it once.

I'd suggest to move these operations our of this loop and into a separate function for config processing. We can also have the validation there later on.

File: pkg/app/zipper/tldcache.go
Comment (grzkv) [inline pkg/app/zipper/tldcache.go:78]: I'm not sure we need to do this every time. As mentioned in another comment, this could be moved to a separate function to process prefixes config.

File: pkg/app/zipper/tldcache.go
Comment (grzkv) [inline pkg/app/zipper/tldcache.go:19]: I think we may consider storing the list of prefixes as a parsed, trimmed, and sorted tree of strings (it will more trie actually). The operations will become much simpler. The code will be more clear. We won't need to do processing repeatedly. The performance will be better. There will be less memory allocation. More importantly, the validation will be backed-in.

@emadolsky What do you think?

**Review by grzkv — DISMISSED** [review]
I have left a number of comments. We can merge this as-is and address them later.
So, approving this for now to unblock even though there are still questions pending.
@emadolsky I'm happy to discuss the questions now if you think that's more appropriate.

File: pkg/cfg/common.go
Comment (emadolsky) [reply pkg/cfg/common.go:182]: Yes it can be implemented like this. however, I don't see a huge benefit from implementing it this way.
Also, it'll create and ambiguity that when we configure:
```tldCacheExtraPrefixes:
  - "a":
      "b":
        - "c"
        - "d"
  - "hello"```
will it induce probing of `a` or `a.b` or is it only `a.b.c` and `a.b.d`.
Although it might not be a normal use-case, probing `a.b.c` should not necessarily cause probing of `a.b`. Of course we can add some attributes to each prefix to identify if it should be probed, but I believe that is not neat.

File: pkg/app/zipper/tldcache.go
Comment (emadolsky) [reply pkg/app/zipper/tldcache.go:20]: Yes I agree. invalid inputs probably do not cause big harms but still it's nice to have. I was even thinking about keeping and using a validated sorted list of prefixes so we wouldn't need to sort it every time. Also it'll make the query lookups faster. 
We can either do it now or later. However, if I do it right now, it's likely that it should be moved and changed during your refactor. So, your call.

File: pkg/app/zipper/tldcache.go
Comment (emadolsky) [reply pkg/app/zipper/tldcache.go:22]: Yeah exactly. I described the same thing on a previous comment.

File: pkg/app/zipper/tldcache.go
Comment (emadolsky) [reply pkg/app/zipper/tldcache.go]: I already made the changes. please check again - @grzkv.

File: pkg/app/zipper/tldcache.go
Comment (emadolsky) [reply pkg/app/zipper/tldcache.go:59]: It's the name of the function. Capitalizing it, make it wrong.

Comment (timtofan) [reply pkg/app/zipper/tldcache.go:59]: I meant "It reuses .." and "If no info is there .." sentences.

Comment (emadolsky) [reply pkg/app/zipper/tldcache.go]: Oh sorry. Fixed.

File: pkg/cfg/common.go
Comment (grzkv) [reply pkg/cfg/common.go:182]: @emadolsky Yes, there are pros and cons to both. I'm totally ok if we leave this like it is.

**Review by grzkv — APPROVED** [review]

File: pkg/app/zipper/http_handlers.go
Comment (timtofan) [inline pkg/app/zipper/http_handlers.go:572]: nit: we are using both Tld and TopLevelDomain in same code, imho we could standardize on one of the two for consistency, maybe Tld since it's used also in package name and config_

File: pkg/app/zipper/tldcache.go
Comment (timtofan) [inline pkg/app/zipper/tldcache.go:102]: might be worth to move this into a func, and add tests for it

Comment (timtofan) [reply pkg/app/zipper/tldcache.go:102]: also IIUC validation could be also done via strings.Contains(p,"..")

**Pull request merged**
